File: src/main/java/io/github/takenoha/towerdefense/persistence/TowerUpgradeResult.java
package io.github.takenoha.towerdefense.persistence;

import java.util.Objects;
import java.util.Optional;

/** Result of applying one idempotent tower upgrade operation. */
public record TowerUpgradeResult(
        OperationOutcome outcome,
        Optional<TowerRecord> tower) {
    public TowerUpgradeResult {
        Objects.requireNonNull(outcome, "outcome");
        tower = Objects.requireNonNull(tower, "tower");
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/persistence/TowerUpgradeResult.kt
package io.github.takenoha.towerdefense.persistence

import java.util.Objects
import java.util.Optional
import kotlin.jvm.JvmRecord

/** Result of applying one idempotent tower upgrade operation. */
@JvmRecord
data class TowerUpgradeResult(
    val outcome: OperationOutcome,
    val tower: Optional<TowerRecord>,
) {
    init {
        Objects.requireNonNull(outcome, "outcome")
        Objects.requireNonNull(tower, "tower")
    }
}
